Have compute tags in GeneralRelativity use mutating functions 2/2

carmaza opened the pull request:
## Proposed changes

Finish up what was started in #2107 

### Types of changes:

- [ ] Bugfix
- [x ] New feature
- [ ] Refactor

### Component:

- [x ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

carmaza committed:
Add mutating overload for inverse spacetime metric

carmaza committed:
Add mutating overload for derivatives_of_spacetime_metric

carmaza committed:
Add mutating overload for spacetime_normal_one_form

carmaza committed:
Add mutating overload for extrinsic_curvature

carmaza committed:
Have compute tags of spacetime quantities use mutating overload

carmaza committed:
Have GH tags' function members return by reference

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:
```diff
@@ -841,38 +842,49 @@ template <size_t SpatialDim, typename Frame>
+      const gsl::not_null<Scalar<DataVector>*> gamma1,
       const tnsr::I<DataVector, SpatialDim, Frame>& coords) noexcept {
-    return make_with_value<type>(coords, -1.);
+    destructive_resize_components(gamma1, get<0>(coords).size());
```
You inconsistently do the destructive_resize in the hpps. Could you use it in the other compute tags that have the function defined in line?

carmaza replied:
I added it to all functions. I thought the assignment in the function implementation would make the resize automatically? If not, I'm afraid there are several other functions in GR/GH that might require the manual `destructive_resize` :P

nilsdeppe replied:
Yes, the assignment does the check. Either is fine, though it would be nice for them to (eventually) be consistent :)

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp:
```diff
@@ -506,22 +513,24 @@ struct DerivSpacetimeMetricCompute
+          deriv_spacetime_metric,
       const tnsr::abb<DataVector, SpatialDim, Frame>&
           spacetime_deriv_of_spacetime_metric) noexcept {
-    auto deriv_spacetime_metric =
```
destructive_resize

nilsdeppe commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp:
```diff
@@ -535,9 +544,14 @@ template <size_t SpatialDim, typename Frame, typename DataType>
+
+  static void function(const gsl::not_null<Scalar<DataType>*> result,
+                       const Scalar<DataType>& det_spatial_metric) noexcept {
+    get(*result) = sqrt(get(det_spatial_metric));
```
destructive_resize

nilsdeppe reviewed (changes requested):
I'm happy for you to squash immediately :)

kidder commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.cpp:
```diff
@@ -227,38 +226,61 @@ tnsr::abb<DataType, Dim, Frame> derivatives_of_spacetime_metric(
+    const tnsr::ijj<DataType, Dim, Frame>& deriv_spatial_metric) noexcept {
+  destructive_resize_components(spacetime_deriv_spacetime_metric,
+                                get_size(get(lapse)));
+  for (auto& component : *spacetime_deriv_spacetime_metric) {
```
looks like the only components that need to be zero initially are `spacetime_deriv_spacetime_metric(a,0,i)` where `a` is all dim+1 spacetime components, and `i` is just the dim spatial components

kidder commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.cpp:
```diff
@@ -346,31 +346,47 @@ void spacetime_normal_vector(
         deriv_spatial_metric) noexcept {
-  const DataType half_over_lapse = 0.5 / get(lapse);
+  destructive_resize_components(ex_curvature, get_size(get(lapse)));
+  for (auto& component : *ex_curvature) {
```
Do we need to initialize this to zero, if we reorder the operations below to first set the `ex_curvature` to `-dt_spatial_metric`?

kidder commented on src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp:
```diff
@@ -930,21 +944,29 @@ struct SpacetimeDerivGaugeHCompute : SpacetimeDerivGaugeH<SpatialDim, Frame>,
+    destructive_resize_components(spacetime_deriv_gauge_source,
+                                  get<0>(time_deriv_gauge_source).size());
+    for (auto& component : *spacetime_deriv_gauge_source) {
+      component = 0.0;
```
I don't think you need this as all components are set explicitly

kidder commented on src/Evolution/Systems/GeneralizedHarmonic/Initialize.hpp:
```diff
@@ -188,10 +188,12 @@ struct InitializeGauge {
-        GeneralizedHarmonic::Tags::SpacetimeDerivGaugeHCompute<
-            Dim, frame>::function(std::move(dt_initial_gauge_source),
-                                  std::move(d_initial_gauge_source));
+    typename GeneralizedHarmonic::Tags::SpacetimeDerivGaugeHCompute<
```
this type should just be `tnsr::ab<DataVector, Dim, frame>`

carmaza commented:
@kidder I implemented all of your change requests and squashed immediately. I wasn't paying too much attention to the actual implementation of the functions in these files, but I'd be happy to do so as part of a separate commit/PR if you think it would help with performance.

kidder commented:
Thanks @carmaza !  I just figure I'd comment on things I noticed.  If you want to look for further improvements in a future PR that would be fine, but I would put it at a lower priority than converting compute items to mutating, and using known types instead of (const_)item_type...
